### ISSUE TITLE
[M12] DATA-8: the tool-calling and reasoning SFT corpora have no training driver

### ISSUE THREAD

**Issue by travisgalloway** [travisgalloway/monica#306]
Found by `/closure-audit` (2026-08-18), pass A. Capability row `DATA-8` in
`docs/feature-matrix.md`.

## Capability
A person can run an SFT pass over the tool-calling and reasoning corpora this repo already knows
how to build — without hand-converting them into the generic format first.

## The gap

Two corpus builders exist, are fully tested, and have **no non-test consumer**:

- `src/data/tool_sft.py` — `build_tool_sft` plus a `__main__` CLI (`:142,229,254`). Only
  `tests/test_tool_sft.py` touches it.
- `src/data/reasoning_sft.py` — `build_reasoning_sft` (`:73,167,195`). Only
  `tests/test_reasoning_sft.py` and `tests/test_storage.py` touch it.

`scripts/sft.py` reads only generic `data/sft/{train,val}.jsonl`; it has no handling for the
`reasoning-packed/` form or for tool-call rows. So both builders produce output nothing consumes.

This is self-documented in two places, which is how it stayed invisible:

- `docs/reserve/path-b-run.md:239-242` — *"`src/data/reasoning_sft.py` builds the data, but there
  is **no dedicated reasoning-SFT training driver** in `scripts/`"*.
- `docs/design/11-post-training.md:56-62` — a **live, non-reserve** doc describing the two output
  forms as though they are usable.

#102 ("Tool-use SFT") was closed COMPLETED with only the builder landed. #101 does not cover this:
its body is M10/#65 framing about an instruct layer and a conversion teacher, not these corpora
(see the comment left on #101).

## Scope

This is one vertical slice, not two, because both failures are the same seam: `scripts/sft.py`
cannot consume a corpus form that is not the generic one.

In scope:
- Teach `scripts/sft.py` to consume the tool-call and reasoning-packed forms the builders emit —
  a corpus-form argument or an autodetect from the manifest, whichever matches the existing
  loader conventions.
- Keep response masking correct for each form: the assistant span for reasoning traces, the
  tool-call payload for tool rows.
- One end-to-end test per form, from builder output through a training step.

Out of scope:
- New corpus sources, new data, or changing what the builders emit.
- DPO/GRPO over these corpora.
- Re-framing #101, which is its own decision.

## Definition of done
- [x] `scripts/sft.py` trains from `build_tool_sft` output without manual conversion.
- [x] `scripts/sft.py` trains from `build_reasoning_sft` output, including the `reasoning-packed/`
      form named in `docs/design/11-post-training.md`.
- [x] Response masking is correct per form and asserted by a test — a wrong mask trains on the
      prompt and is silent.
- [x] Chat-template EOS stays consistent across SFT, RL and serving; a mismatch degrades the model
      at serving time and is the failure this area is most prone to.
- [x] Both builders now have a real non-test consumer, which is the condition that closes this.
- [x] `docs/design/11-post-training.md` describes how to actually run each form.
- [x] `docs/reserve/path-b-run.md:239-242`'s "no dedicated driver" statement is corrected — it is
      a reserve doc, but that specific sentence becomes false.
- [x] `docs/feature-matrix.md` `DATA-8` moves to `Shipped`; `docs/test-plan.md`'s `DATA-8` Gaps
      cell is rewritten.

## Edge cases
- A reasoning trace longer than the sequence length → truncation must not orphan the answer span.
- A tool row with no tool call, or with several.
- Malformed JSON in a tool-call payload → a named error, not a crash mid-epoch.
- An empty corpus after filtering → fail loudly rather than train on nothing.
- Mixing forms in one run → either supported and tested, or rejected with a clear message.

## Verification
- `.venv/bin/python -m pytest tests/test_tool_sft.py tests/test_reasoning_sft.py tests/test_sft_train_step.py tests/test_sft_loader.py -q`
- `python -m src.data.tool_sft --out <tmp>` then `.venv/bin/python scripts/sft.py --config config/toy.yaml --data <tmp>` completing a step.
- The same for `build_reasoning_sft`.
- `.venv/bin/python -m pytest -q` full suite green.

**Comment by travisgalloway** [comment]
Implemented on `feat/306-sft-training-driver` → PR #319.

`scripts/sft.py` is reused as the single SFT driver (no second trainer); the new `src/data/sft_corpus.py` resolves the `shared/sft/tokenized/<tok>-<k>/` layout to trainable records, verifies `chat_eos`/`template` at load, splits deterministically, and rejects `reasoning-packed` by name with the `split --shards` → `scripts/train.py` recipe (packed carries no loss mask). `tests/test_sft_driver_e2e.py` drives builder output through 20 real optimizer steps + a resume round-trip; `tests/test_sft_corpus.py` covers every loud-failure path. Full suite: 1763 passed, 22 skipped. Smoke gate green.
